### PR TITLE
WIP - Add a lazy layout grid for dynamically sized grid tiles

### PR DESCRIPTION
**Pull request by Piinks**
### WIP / Experimental 

> **Note**
> There are still changes to be made here, and several things are exposed currently (to facilitate testing) that likely will not be exposed in a final proposal.

[Use with this Gist](https://gist.github.com/Piinks/369a8c4762f85dda469bac5f9f51a529)

Related Issues
- https://github.com/flutter/flutter/issues/97544
- https://github.com/flutter/flutter/issues/40856

This is a  foundation for some dynamic grid layout patterns @snat-s and @DavBot02 are working on.
Bifurcates SliverGrid into two different render objects

- **RenderSliverGrid** - maintains the original layout algorithm, which benefits greatly for performance since it already knows the layout pattern
- **RenderDynamicSliverGrid** - very similar to SliverList, which can support grid tiles adapting to the size of the child.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 1]
```diff
@@ -317,6 +399,7 @@ class SliverGridDelegateWithFixedCrossAxisCount extends SliverGridDelegate {
     this.crossAxisSpacing = 0.0,
     this.childAspectRatio = 1.0,
     this.mainAxisExtent,
+    super.layoutType,
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:402]: @DavBot02 can you check your branch to make sure you have this change?

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 2]
```diff
@@ -416,6 +500,7 @@ class SliverGridDelegateWithMaxCrossAxisExtent extends SliverGridDelegate {
...
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:503]: @DavBot02 here as well

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 3]
```diff
@@ -105,17 +126,77 @@ class SliverGridGeometry {
...
-  int getMinChildIndexForScrollOffset(double scrollOffset);
+  /// TODO(Piinks): Add docs
+  final SliverGridLayoutType layoutType;
+
+  /// The minimum child index that intersects with (or is after) this scroll
+  /// offset.
+  /// TODO(Piinks): Add more docs - only used for fixed layouts
+  int getMinChildIndexForScrollOffset(double scrollOffset) {
+    return 0;
+    // assert(
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:139]: I've removed several assertions here. I don't know that they are necessary, we can revisit them when we are getting closer to final designs.

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 4]
```diff
@@ -59,11 +59,19 @@ class SliverGridGeometry {
...
+  /// axis. Otherwise, the provided cross axis size or the
+  /// [SliverConstraints.crossAxisExtent] will be used to create tight
+  /// constraints in the cross axis.
   BoxConstraints getBoxConstraints(SliverConstraints constraints) {
     return constraints.asBoxConstraints(
-      minExtent: mainAxisExtent,
+      minExtent: mainAxisExtent.isFinite ? mainAxisExtent : 0,
       maxExtent: mainAxisExtent,
       crossAxisExtent: crossAxisExtent,
     );
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:77]: This has been updated, you can copy paste the function into your branch. Probably simpler than merging if there are merge conflicts.

File: packages/flutter/lib/src/rendering/sliver.dart [hunk 5]
```diff
@@ -369,32 +369,42 @@ class SliverConstraints extends Constraints {
...
-  /// constraint in the cross axis. Otherwise, the [crossAxisExtent] from this
-  /// object is used as a constraint in the cross axis.
+  /// axis. If non-null, and the given `crossAxisExtent` is finite, it is used
+  /// as a tight constraint in the cross axis. When [double.infinity], the
+  /// constraints will be loose in the cross axis. If the provided
+  /// `crossAxisExtent` is null, the [crossAxisExtent] from this object is used
+  /// as a tight constraint in the cross axis.
   ///
   /// Useful for slivers that have [RenderBox] children.
   BoxConstraints asBoxConstraints({
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver.dart:382]: The changes to this function are also new here. You will want to update your branches with this as well.

**Comment by goderbauer** [comment]
Closing for now.

**Pull request closed without merging**